Write NFRs and combine with code standards (#3824)
bbc/simorgh: docs/Code-Standards.md

- Smaller PRs are better, and it's much better for visibility if you open your PRs as draft PRs as early as possible.

## Automated Processes

- Linting of our code: A combination of the [Airbnb Styleguide](https://github.com/airbnb/javascript/tree/master/react) and [Prettier](https://github.com/prettier/prettier) is used as our linting standards.
- Issue and PR templates should be followed, then refined by the relevant internal BBC team. We have no templates for hot-fixes which are what we call live and default branch patches; in such cases maximum speed of work is the priority.
- Most PRs will require code owner approval. This requirement and its enforcement is entirely automated. We have code maintainers across our repos ([Simorgh](https://github.com/bbc/simorgh)/[Psammead](https://github.com/bbc/psammead)), who are responsible for ensuring the quality of the code and are a good source for enquiries.
- AMP validation for pages
- BBC a11y validation for pages
- Lighthouse validation
- Code Climate Quality tests

## Manual Processes

- Open your PRs as early as possible, as draft PRs.

> - Add your issues and PRs to the project "Simorgh".

Reviewer: Worth mentioning that they should add the relevant label for whichever stream of work it is for?

Author: Yes please write that as a suggestion @amywalker1987

Author: Done

> - Add your issues and PRs to the project "Simorgh".

Reviewer: ```suggestion
- Add your issues and PRs to the project "Simorgh".
- Add labels to your issues and PRs to identify which pod/stream of work they lie with. e.g "simorgh-core-stream"
```

Author: Decided to rewrite this a little

- If you have paired on a PR you should co-author commits.
- If latest has a broken build, stop what you are doing and fix it.
- Code reviews take precedence over implementation.

> - Accessibility swarms and UX reviews are required for all front-end components - although they are optional for components with alpha version numbers.

Reviewer 2: Maybe something should be said about using alpha components in simorgh

Author: I think not using them in simorgh should be pretty obvious, but I have another issue for documenting how we use them anyhow: https://github.com/bbc/psammead/issues/1838

- Refinement of GitHub issues should be done by the most relevant BBC team.

## Merging guidelines and Non-Functional Requirements (NFRs) for release

### Before Merge

#### Unit Testing:
- Unit testing is a mixture of Jest snapshot testing and enzyme assertion testing.
- We push automated testing as far down the testing pyramid as possible. This means that if something can be tested via unit tests, to ensure we have a test coverage of > 80% on this repo.

> - For strictly **presentational** components Snapshots tests must be present and assertion testing is desirable.

Reviewer 2: >strictly **presentational** components
Should be in psammead surely?

Author: this is intended to be one doc for both repos

Author: Will make that clearer in https://github.com/bbc/psammead/issues/1838

- For **containers** or otherwise logic handling code snapshot testing is not desirable and assertion testing is mandatory.
- A11y tests/swarms (this would be ideal but not required at this stage)
#### Documentation:
- Any documentation updates and new documentation for a given page type or feature should be added in the PR that makes the changes.

> #### E2E Testing:

Reviewer 2: Worth mentioning that e2e's are mandatory for all simorgh routes, and then something about what is smoke tested

Author: @dr3 done

- E2E service config added for a new service/updated for an existing service
- E2E should be updated and running against localhost and disabled for test and live

> ### Only applicable to BBC staff (inline references and links might not be accessible to non-BBC staff)

Reviewer 3: ```suggestion
### Only applicable to BBC staff
```

#### Promoting to test (www.test.bbc.com):
- Manually check all requirements for merge have been met.
- Ensure any new page type is being tested by a11y and lighthouse.
- Any new data endpoints need to be available on test (speak to core pod about enabling these).

> - Raise an issue in the [simorgh-infrastructure](https://github.com/bbc/simorgh-infrastructure/) repo requesting the STM change for going to test

Reviewer 3: ```suggestion
- Raise an issue to request the traffic manager (STM) change for going to test in the [simorgh-infrastructure repo](https://github.com/bbc/simorgh-infrastructure/). Note, this is a private repo only accessible to BBC staff. 
```
